### PR TITLE
self-sign console bug

### LINKED ISSUE #307: x509: certificate signed by unknown authority in sync pod
in the sync pod, @sozercan saw:

```
time="2018-08-30T06:14:21Z" level=info msg="Sync process started"
time="2018-08-30T06:14:22Z" level=info msg="found account fnr5uugfvzt6vyxwjsff0ebb"
time="2018-08-30T06:14:23Z" level=info msg="Error while syncing: cannot sync cluster config: Get https://sertacosatest29.eastus.cloudapp.azure.com/healthz: x509: certificate signed by unknown authority"
```

### PR DESCRIPTION
**Pull request by mjudeikis**
This should do. testing now if we don't have any fallouts from this 

cc: @jim-minter 

fixes #307

### CODE REVIEW

File: cmd/createorupdate/createorupdate.go [hunk 1]
```diff
@@ -141,16 +141,6 @@ func enrich(cs *acsapi.OpenShiftManagedCluster) error {
...
 	}
 
-	// TODO: MS to provide FQDN values for OCP cluster to use.
-	// cs.Properties.FQDN = random-cluster-master-prefix.eastus.cloudapp.azure.com
-	// cs.Properties.OrchestratorProfile.OpenShiftConfig.RouterProfiles[0].FQDN = random-cluster-router-prefix.eastus.cloudapp.azure.com
-	// We will:
-	//    extract random-cluster-master-prefix and put it into ARM to use this for public-ip
-	//    extract random-cluster-router-prefix and put it into service annotation so cloud-plugin could claim DNS name when provisioning IP
-
-	cs.Properties.FQDN = fmt.Sprintf("%s.%s.cloudapp.azure.com", cs.Properties.AzProfile.ResourceGroup, cs.Location)
```
Comment (jim-minter) [inline cmd/createorupdate/createorupdate.go]: Properties.FQDN must be set incoming - please can you update the validation to ensure that it is checked regardless of the externalOnly flag?

File: cmd/createorupdate/createorupdate.go [hunk 2]
```diff
@@ -141,16 +141,6 @@ func enrich(cs *acsapi.OpenShiftManagedCluster) error {
...
-	cs.Properties.OrchestratorProfile.OpenShiftConfig.RouterProfiles[0].FQDN = fmt.Sprintf("%s-router.%s.cloudapp.azure.com", cs.Properties.AzProfile.ResourceGroup, cs.Location)
```
Comment (jim-minter) [inline cmd/createorupdate/createorupdate.go]: this is still set in enrich, I think

File: hack/create.sh [hunk 3]
```diff
@@ -72,6 +73,7 @@ properties:
   routerProfiles:
   - name: default
     publicSubdomain: $RESOURCEGROUP.$DNS_DOMAIN
+    fqdn: $RESOURCEGROUP-router.eastus.cloudapp.azure.com
```
Comment (jim-minter) [inline hack/create.sh]: think that should be reverted?

File: pkg/config/generate.go [hunk 4]
```diff
@@ -211,6 +211,7 @@ func Generate(cs *acsapi.OpenShiftManagedCluster) (err error) {
 		// Openshift Console is BYO type of certificate. In the long run we should
 		// enable users to configure their own certificates.
 		// For this reason we decouple it from all OCP certs and make it self-sign
+		// If FQDN matches PublicHostname certificate can't be self-sign
```
Comment (jim-minter) [inline pkg/config/generate.go]: bit more detail here would be good (or link to issue)

File: cmd/createorupdate/createorupdate.go (hunk 2 above)
Comment (mjudeikis) [reply cmd/createorupdate/createorupdate.go]: Sure? It should be the same incoming from MS too, no?

File: cmd/createorupdate/createorupdate.go (hunk 1 above)
Comment (jim-minter) [reply cmd/createorupdate/createorupdate.go]: Also please can you update the comment about the direction of FQDN in v1/types.go ?

File: pkg/validate/validate.go [hunk 5]
```diff
@@ -221,8 +219,13 @@ func validateFQDN(p *api.Properties) (errs []error) {
...
 	}
-	if !isValidHostname(p.FQDN) {
-		errs = append(errs, fmt.Errorf("invalid properties.fqdn %q", p.FQDN))
+	// check if FQDN is not empty, and validate format if not.
+	if p.FQDN == "" {
+		errs = append(errs, fmt.Errorf("properties.fqdn cannot be empty"))
+	} else {
+		if !isValidHostname(p.FQDN) {
+			errs = append(errs, fmt.Errorf("invalid properties.fqdn %q", p.FQDN))
+		}
```
Comment (mjudeikis) [inline pkg/validate/validate.go]: This prevents both validation errors. `""` is one errors `()` is other.

File: pkg/config/generate.go (hunk 4 above)
Comment (jim-minter) [reply pkg/config/generate.go]: sorry, meant link to https://github.com/openshift/openshift-azure/issues/307

File: pkg/config/generate.go [hunk 6]
```diff
@@ -219,7 +221,7 @@ func Generate(cs *acsapi.OpenShiftManagedCluster) (err error) {
 			extKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
 			key:         &c.Certificates.OpenshiftConsole.Key,
 			cert:        &c.Certificates.OpenshiftConsole.Cert,
-			selfSign:    true,
+			selfSign:    cs.Properties.FQDN != cs.Properties.OrchestratorProfile.OpenShiftConfig.PublicHostname,
```
Comment (jim-minter) [inline pkg/config/generate.go]: for now, how about always sign it with CA? entirely remove the selfSign pieces here and place a comment in the code that when we allow the user to specify this cert, remember issue 307 and they can only specify a cert if PublicHostname != FQDN

File: pkg/config/generate.go (hunk 4 above)
Comment (mjudeikis) [reply pkg/config/generate.go]: Its expected behavior of named_certificates for the console. The issue is a consequence of me failing to implement it the first time. I think the issue will become irrelevant soon?

File: pkg/validate/validate.go (hunk 5 above)
Comment (jim-minter) [reply pkg/validate/validate.go]: simplify to `if p.FQDN == "" || !isValidHostname(p.FQDN) {`

File: pkg/validate/validate.go [hunk 7]
```diff
@@ -77,9 +77,7 @@ func validateUpdateContainerService(cs, oldCs *api.OpenShiftManagedCluster, exte
 func validateProperties(p *api.Properties, externalOnly bool) (errs []error) {
 	errs = append(errs, validateProvisioningState(p.ProvisioningState)...)
 	errs = append(errs, validateOrchestratorProfile(p.OrchestratorProfile, externalOnly)...)
-	if !externalOnly {
-		errs = append(errs, validateFQDN(p)...)
-	}
+	errs = append(errs, validateFQDN(p)...)
```
Comment (jim-minter) [inline pkg/validate/validate.go]: you won't need externalOnly as an argument to validateProperties any more - can you remove it from the functions where it's not needed?

Comment (mjudeikis) [reply pkg/validate/validate.go]: `errs = append(errs, validateOrchestratorProfile(p.OrchestratorProfile, externalOnly)...)` for router profile?

Comment (mjudeikis) [reply pkg/validate/validate.go]: Or we dont want to validate this via `externalOnly` too? Missing logic in my head for this

Comment (jim-minter) [reply pkg/validate/validate.go]: bah, ignore me, sorry

**Comment by jim-minter** [comment]
please squash and will lgtm

**Pull request merged**